rfcs/0028-nix-release-model.md

```diff
@@ -0,0 +1,163 @@
+---
+feature: nix-release-model
+start-date: 2018-03-28
+author: Shea Levy <shea@shealevy.com>
+co-authors: Michael Raskin <7c6f434c@mail.ru> (any other Nix committers on-board?)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+A model of when, what, and how to make new releases of Nix.
+
+# Motivation
+[motivation]: #motivation
+
+We want to set clear expectations for users, provide guidance for
+developers, and provide both groups the benefits of easy availability
+of new improvements.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed process contains a number of interrelated components.
+I've numbered them for ease of reference and discussion, not to
+indicate any meaningful ordering. Any aspect of this can be suspended
+at the security team's discretion when handling a vulnerability, or at
+the core team's discretion for especially destructive bugs (e.g. data
+corruption).
+
+1. Before merging to `master`, require testing, documentation,
+   changelog entries, removal of WIP markers, etc. to keep `master`
+   somewhat stable, maintain bisectability, and shorten the release
+   preparation process. This includes both objective and subjective
+   components, and will always involve something of a judgment call.
+2. The Nix core team shall decide when a new release is warranted,
+   through its normal decision procedures. That decision will also
+   include designating a release manager for the release. The specific
+   conditions for a release being warranted are intentionally out of
+   scope of this RFC, but it is expected that the core team will
+   balance both the needs of the community at large and the needs of
+   developers working on Nix and the tradeoff between stability and
+   development progress. In particular, if an especially major change
+   has landed recently or is nearing readiness, the core team should
+   either mark the release point as being before the relevant change
+   or expect more stabilization work than normal.
+3. The release manager will branch off a `#-maintenance` branch
+   (e.g. `2.1-maintenance`) in advance of any mainline release. A
+   mainline release is a release that is a direct git ancestor of
+   `master` and is, in the relevant feature and bugfix senses, "ahead"
+   of all other releases at the time it is made.
```

Comment:
If some stabilisation is needed the release itself will not be a direct git ancestor of master.

Reply:
Ah, my claim also assumes some minor feature development on master, but that can happen, too…

Reply:
It will be, because we merge forward during stabilization

Reply:
So basically 
> No commit should exist on the `#-maintenance` branch during the stabilization phase without very soon thereafter being merged into `master` except under rare circumstances.
means just that «normally, maintenance branch should be merged into master frequently»?

Reply:
Yep... what else would it mean?

Reply:
Well, when you focus on commits being merged, there are interesting things that can happen with commits. Like a branch with a couple of fixes being merged into both maintenance and master.

Reply:
> Changes aimed at getting the release ready (fixing bugs, improving docs, etc.) should target the #-maintenance branch and then be merged forward into master.

This seems clear enough to me. If someone merges a branch twice instead of merging into the maintenance branch and then merging maintenance into master, that's not the expected workflow but it's also not actually a big deal... So there's a merge commit that they differ by, trivial to fix up with a merge later if we need to.

Reply:
I just think that if you define things like «mainline release» in a way that assumes the expected workflow, and you have clear expectations, maybe a more explicit description is a good idea. It is not even longer anyway.

Reply:
Can you open a PR with your proposed rewording?